lib/rpc/broker.toit

```diff
@@ -98,39 +114,81 @@ monitor RpcRequestQueue_:
+    return true
+
+  remove_first -> RpcRequest_?:
+    while true:
+      request := first_
+      if not request:
+        deadline := Time.monotonic_us + (IDLE_TIME_MS * 1_000)
+        if not (try_await --deadline=deadline: first_ != null):
+          return null
+        continue
+
+      // Dequeue the first request from the linked list.
+      next := request.next
+      if identical last_ request: last_ = next
+      first_ = next
+      return request
+
+  cancel pid/int id/int -> int:
+    // For testing purposes, we keep track of the number of requests that
+    // were actually canceled through this operation.
+    result/int := 0
+    // First we get rid of any unprocessed request with the given id. This
+    // is a simple linked list traversal with the usual bookkeeping challenges
+    // that come from removing from a linked list with insertion at the end.
+    previous := null
+    current := first_
+    while current:
+      next := current.next
+      if current.pid == pid and current.id == id:
+        if previous:
+          previous.next = next
+        else:
+          first_ = next
+        if not next:
+          last_ = previous
+        unprocessed_--
+        result++
+      previous = current
+      current = next
+    // Then we cancel any requests that are in progress by canceling the
+    // associated processing task.
+    processing_requests_.size.repeat:
+      request/RpcRequest_? := processing_requests_[it]
+      if request and request.pid == pid and request.id == id:
+        processing_tasks_[it].cancel
+        result++
+    return result
+
+  ensure_processing_task_ -> none:
     // If there are requests that could be processed by spawning more tasks,
     // we do that now. To avoid spending too much memory on tasks, we prefer
     // to keep some requests unprocessed and enqueued.
-    while unprocessed_ > tasks_ and tasks_ < RpcBroker.MAX_TASKS:
+    while unprocessed_ > tasks_ and tasks_ < processing_tasks_.size:
       tasks_++
-      task --background::
+      task_index := processing_tasks_.index_of null
+      processing_tasks_[task_index] = task --background::
         // The task code runs outside the monitor, so the monitor
         // is unlocked when the requests are being processed but
         // locked when the requests are being dequeued.
+        assert: identical processing_tasks_[task_index] task
         try:
-          while next := remove_first:
+          while not task.is_canceled:
+            next := remove_first
+            if not next: break
             try:
+              processing_requests_[task_index] = next
               next.process
             finally:
               // This doesn't have to be in a finally-block because the call
               // to 'next.process' never unwinds, but being a little bit
               // defensive feels right.
+              processing_requests_[task_index] = null
               unprocessed_--
         finally:
```

Comment:
I suppose it is the same for this `finally`; `process` will return from its `finally` which stops any on going unwinding, so the `finally` is strictly not necessary? I at least used this line of thinking to convince myself that a critical block wasn't necessary to prevent a cancel from stopping `ensure_processing_task_` prematurely.

Reply:
The finally here is more for if `remove_first` or any other stuff outside the inner try-block should error out.